pack: gelf: add string log level support

alitoufighi opened the pull request:
As [GELF Payload Specification](https://docs.graylog.org/en/latest/pages/gelf.html#gelf-payload-specification) mentions, it uses [Syslog Severity Levels](https://en.wikipedia.org/wiki/Syslog#Severity_level). So I thought we can support string values for logs.
Sometimes your applications use different numerical standards for levels, but usually the string representations are the same. I picked up mapped values from link above, which refers to Linux manpage for `syslog`.

edsiper commented:
@manuelluis if I am not wrong you wrote the Gelf part, can you review this ?

manuelluis commented on src/flb_pack.c:
```diff
@@ -1512,9 +1512,31 @@ flb_sds_t flb_msgpack_to_gelf(flb_sds_t *s, msgpack_object *o,
+                        int i;
+                        
+                        for (i = 0; i < array_size; ++i) {
+                            if (!strncmp(mk_string_tolower(val), allowed_levels[i], val_len)) {
```
The `mk_string_tolower` does a malloc using `mk_string_dup`, so we are leaking memory, and if I'm not wrong msgpack strings are not null terminated. It's best to use `strncasecmp`

alitoufighi replied:
Agreed. Changed to `strncasecmp`. I also changed the size of comparison to `allowed_levels[i]` to make it more accurate.
Anyway, this also matches `errooooo`. IMO, it is okay, because I think it's better to be optimistic about logs and ship them as long as we can. But if you think it needs to be exact, I can also check equality of length of strings.

manuelluis commented on src/flb_pack.c:
```diff
@@ -1512,9 +1512,31 @@ flb_sds_t flb_msgpack_to_gelf(flb_sds_t *s, msgpack_object *o,
+                        
+                        for (i = 0; i < array_size; ++i) {
+                            if (!strncmp(mk_string_tolower(val), allowed_levels[i], val_len)) {
+                                v->type = MSGPACK_OBJECT_POSITIVE_INTEGER;
+                                v->via.u64 = (uint64_t)i;
```
I think it is not a good idea to change the `msgpack_object` modifying the type , because if we read it later, strange things can happen, it lost the size of the string.

manuelluis replied:
I think that is better to use another `msgpack_object` for the level number:

```
diff --git a/src/flb_pack_gelf.c b/src/flb_pack_gelf.c
index 96c1980c..90c9dc05 100644
--- a/src/flb_pack_gelf.c
+++ b/src/flb_pack_gelf.c
@@ -496,6 +496,7 @@ flb_sds_t flb_msgpack_to_gelf(flb_sds_t *s, msgpack_object *o,
 
             msgpack_object *k = &p[i].key;
             msgpack_object *v = &p[i].val;
+            msgpack_object vtmp;
 
             if (k->type != MSGPACK_OBJECT_BIN && k->type != MSGPACK_OBJECT_STR) {
                 continue;
@@ -555,10 +556,27 @@ flb_sds_t flb_msgpack_to_gelf(flb_sds_t *s, msgpack_object *o,
                 else if (v->type == MSGPACK_OBJECT_STR) {
                     val     = v->via.str.ptr;
                     val_len = v->via.str.size;
-                    if (val_len != 1 || val[0] < '0' || val[0] > '7') {
-                        flb_error("[flb_msgpack_to_gelf] level is '%.*s', "
-                                  "but should be in 0..7", val_len, val);
-                        return NULL;
+                    if (!(val_len == 1 && val[0] >= '0' && val[0] <= '7')) {
+                        int n;
+                        char* allowed_levels[] = {
+                            "emerg", "alert", "crit", "err",
+                            "warning", "notice", "info", "debug",
+                            NULL
+                        };
+                        for (n=0; allowed_levels[n] != NULL; n++) {
+                            if (!strncasecmp(val, allowed_levels[n], val_len)) {
+                                v = &vtmp;
+                                v->type = MSGPACK_OBJECT_POSITIVE_INTEGER;
+                                v->via.u64 = (uint64_t)n;
+                                break;
+                            }
+                        }
+                        if (allowed_levels[n] == NULL) {
+                            flb_error("[flb_msgpack_to_gelf] level is '%.*s', "
+                                      "but should be in 0..7 or a syslog level",
+                                       val_len, val);
+                            return NULL;
+                        }
                     }
                 }
             }
```

alitoufighi replied:
I modified my code, but aren't we going to override the old content by rewriting the value of union?
I checked the later assignment of value:
```
else if (v->type == MSGPACK_OBJECT_POSITIVE_INTEGER) {
    val = temp;
    val_len = snprintf(temp, sizeof(temp) - 1,
                        "%" PRIu64, v->via.u64);
}
```
And I don't think this causes any problems.
If you think we can change this back, I think it'd be better to do.

alitoufighi replied:
@manuelluis Would you please give your feedback about this, too?

manuelluis replied:
I don't know if I understood your question, when we do:
```
v = &vtmp;
v->type = MSGPACK_OBJECT_POSITIVE_INTEGER;
v->via.u64 = (uint64_t)n;
```
where are point the `msgpack_object *v` to vtmp that is a struct in the stack, so there  is no allocation in the heap and `v` are not pointing to the `msgpack_object` that we get in the args.

and in:
```
else if (v->type == MSGPACK_OBJECT_POSITIVE_INTEGER) {
    val = temp;
    val_len = snprintf(temp, sizeof(temp) - 1,
                        "%" PRIu64, v->via.u64);
}
```
we are writing over `temp` that is  a char of 48 bytes (enough to write a 64 bit integer) initialized to 0 in the stack, we are not modifying the `msgpack_object`.

alitoufighi replied:
Yeah, I missed the point that we lose the pointer to that allocated string. You're right.

manuelluis commented:
Also for fix: https://github.com/fluent/fluent-bit/issues/1717 convert strings "0" to "7" to numbers:

```
diff --git a/src/flb_pack_gelf.c b/src/flb_pack_gelf.c
index 96c1980c..977e16f4 100644
--- a/src/flb_pack_gelf.c
+++ b/src/flb_pack_gelf.c
@@ -496,6 +496,7 @@ flb_sds_t flb_msgpack_to_gelf(flb_sds_t *s, msgpack_object *o,
 
             msgpack_object *k = &p[i].key;
             msgpack_object *v = &p[i].val;
+            msgpack_object vtmp;
 
             if (k->type != MSGPACK_OBJECT_BIN && k->type != MSGPACK_OBJECT_STR) {
                 continue;
@@ -555,10 +556,32 @@ flb_sds_t flb_msgpack_to_gelf(flb_sds_t *s, msgpack_object *o,
                 else if (v->type == MSGPACK_OBJECT_STR) {
                     val     = v->via.str.ptr;
                     val_len = v->via.str.size;
-                    if (val_len != 1 || val[0] < '0' || val[0] > '7') {
-                        flb_error("[flb_msgpack_to_gelf] level is '%.*s', "
-                                  "but should be in 0..7", val_len, val);
-                        return NULL;
+                    if (val_len == 1 && val[0] >= '0' && val[0] <= '7') {
+                        v = &vtmp;
+                        v->type = MSGPACK_OBJECT_POSITIVE_INTEGER;
+                        v->via.u64 = (uint64_t)(val[0] - '0');
+                    }
+                    else {
+                        int n;
+                        char* allowed_levels[] = {
+                            "emerg", "alert", "crit", "err",
+                            "warning", "notice", "info", "debug",
+                            NULL
+                        };
+                        for (n=0; allowed_levels[n] != NULL; n++) {
+                            if (!strncasecmp(val, allowed_levels[n], val_len)) {
+                                v = &vtmp;
+                                v->type = MSGPACK_OBJECT_POSITIVE_INTEGER;
+                                v->via.u64 = (uint64_t)n;
+                                break;
+                            }
+                        }
+                        if (allowed_levels[n] == NULL) {
+                            flb_error("[flb_msgpack_to_gelf] level is '%.*s', "
+                                      "but should be in 0..7 or a syslog level",
+                                       val_len, val);
+                            return NULL;
+                        }
                     }
                 }
             }
```

alitoufighi commented:
Hi all. It was fun and also strange to see finally a one-year old PR gets reviewed :D 
Here are the changes:
* Moved the changes to new file (`flb_pack_gelf.c`).
* Used a separate `msgpack_object` when changing type to prevent unexpected results.
* Added that clients can use syslog level strings to log.
* Fixed https://github.com/fluent/fluent-bit/issues/1717.

manuelluis commented on src/flb_pack_gelf.c:
```diff
@@ -548,17 +549,38 @@ flb_sds_t flb_msgpack_to_gelf(flb_sds_t *s, msgpack_object *o,
+                        v->via.u64 = (uint64_t)(val[0] - '0');
+                    }
+                    else {
+                        int i;
```
It's confusing, at least for me, use the same variable of the upper loop

alitoufighi replied:
Ah, I didn't mention the upper level `i`. Now it's confusing for me too!

manuelluis commented on src/flb_pack_gelf.c:
```diff
@@ -548,17 +549,38 @@ flb_sds_t flb_msgpack_to_gelf(flb_sds_t *s, msgpack_object *o,
+                            NULL
+                        };
+                        for (i = 0; allowed_levels[i] != NULL; ++i) {
+                            if (!strncasecmp(val, allowed_levels[i], strlen(allowed_levels[i]))) {
```
Since msgpack string is not null terminated the third argument of `strncasecmp` should be val_len, if not you can be reading bytes out of the msgpack buffer.

Also you are right, we need to check the length of the strings:
```
if ((val_len == strlen(allowed_levels[i])) && !strncasecmp(val, allowed_levels[i], val_len))
```

alitoufighi replied:
Okay, I'll add this asap.
But just for curiosity, the reason that you mentioned they're not null terminated is about possible page faults, or about security reasons?

manuelluis replied:
Yes, you can get a page fault. 
The buffer over-read it's a security problem, the implications depends on the case.
 https://cwe.mitre.org/data/definitions/126.html

alitoufighi commented:
@edsiper Are we good to go?

edsiper commented:
@manuelluis can you check pls ?

manuelluis commented:
It's ok for me.

edsiper commented:
thanks everyone!, merging...
